build_tools/benchmarks/common/benchmark_suite.py

```diff
@@ -29,16 +29,150 @@
+    driver: the IREE driver to run with, e.g., 'dylib'.
+    benchmark_case_dir: the path to benchmark case directory.
+    benchmark_tool_name: the benchmark tool, e.g., 'iree-benchmark-module'.
+  """
+
+  model_name_with_tags: str
+  bench_mode: str
+  target_arch: str
+  driver: str
+  benchmark_case_dir: str
+  benchmark_tool_name: str
+
+
+class BenchmarkSuite(object):
+  """Represents the benchmarks in benchmark suite directory."""
+
+  def __init__(self, suite_map: Dict[str, List[BenchmarkCase]]):
+    """Construct a benchmark suite.
+    
+    Args:
+      suites: the map of benchmark cases keyed by category directories.
+    """
+    self.suite_map = suite_map
+    self.category_map = dict((os.path.basename(category_dir), category_dir)
+                             for category_dir in self.suite_map.keys())
+
+  def list_categories(self) -> List[Tuple[str, str]]:
+    """Returns all categories and their directories.
+    
+    Returns:
+      A tuple of (category name, category dir).
+    """
+    category_list = [(name, path) for name, path in self.category_map.items()]
+    # Fix the order of category list.
+    category_list.sort(key=lambda category: category[0])
+    return category_list
+
+  def filter_benchmarks_for_category(
+      self,
+      category: str,
+      available_drivers: Sequence[str],
+      cpu_target_arch_filter: str,
+      gpu_target_arch_filter: str,
+      driver_filter: Optional[str] = None,
+      mode_filter: Optional[str] = None,
+      model_name_filter: Optional[str] = None) -> Sequence[BenchmarkCase]:
+    """Filters benchmarks in a specific category for the given device.
+      Args:
+        category: the specific benchmark category.
+        available_drivers: list of drivers supported by the tools.
+        cpu_target_arch_filter: CPU target architecture filter regex.
+        gpu_target_arch_filter: GPU target architecture filter regex.
+        driver_filter: driver filter regex.
+        mode_filter: benchmark mode regex.
+        model_name_filter: model name regex.
+      Returns:
+        A list of matched benchmark cases.
+    """
+
+    category_dir = self.category_map.get(category)
+    if category_dir is None:
+      return []
+
+    chosen_cases = []
+    for benchmark_case in self.suite_map[category_dir]:
+      matched_driver = (benchmark_case.driver in available_drivers) and (
+          driver_filter is None or
+          re.match(driver_filter, benchmark_case.driver) is not None)
+      matched_arch = (re.match(cpu_target_arch_filter,
+                               benchmark_case.target_arch) is not None or
+                      re.match(gpu_target_arch_filter,
+                               benchmark_case.target_arch) is not None)
+      matched_mode = (mode_filter is None or re.match(
+          mode_filter, benchmark_case.bench_mode) is not None)
+
+      # For backward compatibility, model_name_filter matches against the string:
```

Comment:
Not sure if this is intended. If not, I think it is better to just match `<model name-tags>`, e.g. `MobileNetV3Small-fp32,imagenet`